src/storage-operators/src/oneshot_source.rs

```diff
@@ -0,0 +1,880 @@
+// Copyright Materialize, Inc. and contributors. All rights reserved.
+//
+// Use of this software is governed by the Business Source License
+// included in the LICENSE file.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0.
+
+//! "Oneshot" sources are a one-time ingestion of data from an external system, unlike traditional
+//! sources, they __do not__ run continuously. Oneshot sources are generally used for `COPY FROM`
+//! SQL statements.
+//!
+//! The implementation of reading and parsing data is behind the [`OneshotSource`] and
+//! [`OneshotFormat`] traits, respectively. Users looking to add new sources or formats, should
+//! only need to add new implementations for these traits.
+//!
+//! * [`OneshotSource`] is an interface for listing and reading from an external system, e.g. an
+//!   HTTP server.
+//! * [`OneshotFormat`] is an interface for how to parallelize and parse data, e.g. CSV.
+//!
+//! Given a [`OneshotSource`] and a [`OneshotFormat`] we build a dataflow structured like the
+//! following:
+//!
+//! ```text
+//!             ┏━━━━━━━━━━━━━━━┓
+//!             ┃    Discover   ┃
+//!             ┃    objects    ┃
+//!             ┗━━━━━━━┯━━━━━━━┛
+//!           ┌───< Distribute >───┐
+//!           │                    │
+//!     ┏━━━━━v━━━━┓         ┏━━━━━v━━━━┓
+//!     ┃  Split   ┃   ...   ┃  Split   ┃
+//!     ┃  Work 1  ┃         ┃  Work n  ┃
+//!     ┗━━━━━┯━━━━┛         ┗━━━━━┯━━━━┛
+//!           │                    │
+//!           ├───< Distribute >───┤
```

Comment:
We should only distribute once in the beginning, ideally just the descriptions of the objects to fetched, and then run the required work on one worker. Distributing multiple times down the dataflow incurs a network cost and makes the dataflow more susceptible to runaway memory usage since backpressure cannot be communicated cross-worker by default.

Reply:
The reason we distribute multiple times are for the cases where there is a single large file that users want to ingest. For example, a single 200GB Parquet file can be split into hundreds of Row Groups that can be parallelized across any number of workers. Maybe this means we should combine the `Discover` and `Split Work` stages into a single operator, but to keep the implementation as flexible as possible I made them separate.